src/main/scala/io/kaizensolutions/virgil/CQLExecutor.scala

```diff
@@ -0,0 +1,215 @@
+    ZIO.fromCompletionStage(underlyingSession.prepareAsync(query))
+
+  private def executeAction(query: Statement[_]): Task[AsyncResultSet] =
+    ZIO.fromCompletionStage(underlyingSession.executeAsync(query))
+
+  private def select(query: Statement[_]): ZStream[Any, Throwable, Row] = {
+    val initialEffect = ZIO.fromCompletionStage(underlyingSession.executeAsync(query))
+
+    ZStream.fromEffect(initialEffect).flatMap { initial =>
+      ZStream.paginateChunkM(initial) { resultSet =>
+        val emit = Chunk.fromIterable(resultSet.currentPage().asScala)
+        if (resultSet.hasMorePages) {
+          ZIO
+            .fromCompletionStage(resultSet.fetchNextPage())
+            .map(nextState => emit -> Option(nextState))
+        } else {
+          ZIO.succeed(emit -> None)
+        }
+      }
+    }
+  }
+
+  private def selectPage(queryConfiguredWithPageState: Statement[_]): Task[(Chunk[Row], Option[PageState])] =
+    executeAction(queryConfiguredWithPageState).map { rs =>
+      val currentRows = Chunk.fromIterable(rs.currentPage().asScala)
+      if (rs.hasMorePages) {
+        val pageState = PageState.fromDriver(rs.getExecutionInfo.getSafePagingState)
+        (currentRows, Option(pageState))
+      } else (currentRows, None)
+    }
+
+  private def buildStatement(
+    queryString: String,
+    columns: BindMarkers,
+    config: ExecutionAttributes
+  ): Task[BoundStatement] =
+    prepare(queryString).mapEffect { preparedStatement =>
+      val result: BoundStatementBuilder = {
+        val initial = preparedStatement.boundStatementBuilder()
+        val boundColumns = columns.underlying.foldLeft(initial) { case (accBuilder, (colName, column)) =>
+          column.write.write(
+            builder = accBuilder,
+            column = colName.name,
+            value = column.value
+          )
+        }
+        // Configure bound statement with the common execution attributes
+        // https://docs.datastax.com/en/developer/java-driver/4.13/manual/core/statements/
+        config.configure(boundColumns)
+      }
+      result.build()
+    }
+}
+
+object CQLExecutor {
+  def execute[A](in: CQL[A]): ZStream[Has[CQLExecutor], Throwable, A] =
+    ZStream.service[CQLExecutor].flatMap(_.execute(in))
+
+  def executePage[A](in: CQL[A], pageState: Option[PageState] = None)(implicit
+    ev: A =:!= MutationResult
+  ): RIO[Has[CQLExecutor], Paged[A]] =
+    ZIO.serviceWith[CQLExecutor](_.executePage(in, pageState))
+
+  /**
+   * Create a ZIO Cassandra Session from an existing Datastax Java Driver's
+   * CqlSession Note that the user is responsible for the lifecycle of the
+   * underlying CqlSession
+   * @param session
+   *   is the underlying Datastax Java Driver's CqlSession
+   * @return
+   *   the ZIO Cassandra Session
+   */
+  def fromExisting(session: CqlSession): CQLExecutor =
+    new CQLExecutor(session)
+
+  def apply(builder: CqlSessionBuilder): TaskManaged[CQLExecutor] = {
```

Comment:
Consider making a layer here.

Reply:
Added 2, one for the builder and one for an existing session